### DISCUSSION TITLE
Regarding privacy enhancement and coordinations between projects

### DISCUSSION THREAD

**Discussion by wchen342** [Eloston/ungoogled-chromium#1395] [Ideas]
This is a continuation of https://github.com/Eloston/ungoogled-chromium/issues/1393#issuecomment-779825201.

I noticed recently there are increasing demands for enhanced anti-fingerprinting features (e.g. [font spoofing](https://github.com/Eloston/ungoogled-chromium/issues/1393), [favicon fingerprinting](https://github.com/Eloston/ungoogled-chromium/discussions/1388), [enhanced canvas spoofing](https://github.com/Eloston/ungoogled-chromium/issues/1192), [enhanced general fingerprinting resistence](https://github.com/Eloston/ungoogled-chromium/issues/1226)). So far such functionalities are implemented on a case-by-case basis, like the canvas random noise and disabling webrtc. And some of them causes problems, like the canvas random noise (#1241, #1139).

I think it will be much better if these separate enhancements can be somehow grouped into a more systematic project, which will make the issues more concentrated. It will also be better if we can have a coordinated effort between projects on this because a cooperated project can attract more contributions rather than doing this inside a single project. Since Bromite, ungoogled-chromium/ungoogled-chromium-android and probably some other chromium-based forks all have a common goal of enhancing privacy, I think this shall be benifical for all. What do you think? @Eloston @csagan5

**Comment by csagan5** [comment]
I thought the same for the extensions: a project to take care only of that, for the benefit of Chromium forks.

In this case, and possibly some other, I do not think it is necessary some particular form of coordination, what we need is basically:
* [x] (1) use a compatible license
* [ ] (2) redirect request for bugfixes/enhancements to the project that "incubates" and maintains the patchset
* [ ] (3) make each other aware of better patches by contributing back to the upstream of such patchsets, and have some way to inform downstream for the patchset updates

This way contributors can choose which project to contribute to, and then projects can "horizontally" transfer patches like in a sort of software [HGT](https://en.wikipedia.org/wiki/Horizontal_gene_transfer).

For (3) I have often gone back to the source of the patches I use to see if there are updates or if I can contribute back something; that is also a reason to always keep correct attribution. Perhaps we need to identify how we will be upstream/downstream and what will be the channel/way to get the news about the updates, and then (3) would be solved. If you have other ideas please shoot.

I would suggest to start with 1,2,3 and then in future it is possible to say: these patchsets now need their own stand-alone open source project/organisation (either one for all, or one for each), but I would not start now with that - it feels like it would dilute efforts and add bureaucracy.

**Reply by Eloston** [reply]
> Perhaps we need to identify how we will be upstream/downstream and what will be the channel/way to get the news about the updates, and then (3) would be solved.

Also we would need to identify who would be upstream/downstream.

Further, your use of upstream/downstream implies there should be a hierarchical structure for sharing patches & feedback. But you also mentioned HGT, which seems more like a P2P or ad-hoc system to me. Could you clarify the system of sharing you're proposing?

Finally, assuming your approach is more hierarchical, I am still unclear on how this is different enough from setting up another project. Is the bureaucracy you're worried about relating to ownership of features or resolving disagreements on designs/implementations?

**Reply by csagan5** [reply]
> would need to identify who would be upstream/downstream.
> your use of upstream/downstream implies there should be a hierarchical structure for sharing patches & feedback.
> Could you clarify the system of sharing you're proposing?

The only hierarchy is that if you create the patch you are the upstream and maintainer. Everyone else is downstream.

> I am still unclear on how this is different enough from setting up another project. Is the bureaucracy you're worried about relating to ownership of features or resolving disagreements on designs/implementations?

Creating a project requires some work. Branding etc.
I do not see it beneficial to start a project even before creating the patches. You can always spin them off if they have such a critical mass to benefit from a separate project.

**Reply by wchen342** [reply]
I think the problem is rather we lack a clear definition of what is the "enhancement of privacy" we want here, so the efforts are less scattered and less confusing. This is inspired by [the comment](https://github.com/ungoogled-software/ungoogled-chromium-android/issues/63#issuecomment-782770547).

I think at least we need to make clear what level of fingerprint we want for the browser, and whether we shall attempt to mix with the crowd. I still prefer the Tor approach in the long run but it seems there is not enough interest at this point.

**Reply by csagan5** [reply]
@wchen342 in my comment I have outlined a possible way to maintain, share and coordinate work with patches. What is the content of such patches does not matter from the perspective of sharing and collaborating.

> I think at least we need to make clear what level of fingerprint we want for the browser

Anti-fingerprinting is a subset of "privacy enhancement"; this topic started about coordinating the **privacy** patches so I would avoid looking only into anti-fingerprinting now.

Every project has a different level of maturity/knowledge and different definitions; we cannot make definitions converge only by having discussions. What is necessary is to read, learn, share and then discuss together our learnings/experiments. I am sure we all know a lot on these topics, but I expect progress to happen when we add more information to the mix and not just when using what is already in our minds.

Even in the example of GPC, or for other patches, what material are we reading before making code changes? Are we informing ourselves? Why are there no privacy/security researchers involved in these projects? These are the questions I would ask to try to increase the maturity of these projects (all of them, Bromite included).

To go back to @Eloston's question:
> Is the bureaucracy you're worried about relating to ownership of features or resolving disagreements on designs/implementations?

I am worried about introducing or modifying patches based solely on the feedback of a few vocal users; it is a normal occurrence in open source (we use the feedback we have) and quite difficult to counteract. In this sense having different open projects which pick different patches/strategies is a strength; the voices I am missing are the users that do not come to complain, the developers, the researchers...everybody else basically. A browser is a great instrument of freedom, we can probably all agree on that.

@wchen342 to go back to the original question, I can say that my dream has always been to have a "Mozilla for Chromium forks", where by Mozilla I intend the (perhaps too much idealised if compared to the actual one) no-profit organisation.

This could allow at some point even to counterbalance some aspects of the development of Chromium with a stronger voice; but for now we have to go with baby steps, so this is how I can foresee that an improvement from current situation - without creating an absolute centralisation - could be arranged:
1. a shared wiki where we "index" not patches but privacy problems related to Chromium forks; so for example font fingerprinting would be PP-FP-FONT, the WebRTC one could be PP-WEBRTC etc
2. in such wiki we do all the creative work necessary, that is gathering knowledge, links etc the discussions would happen elsewhere (like here via GitHub discussions, or Matrix chat)
3. we document there with some table the corresponding patch or patches used by the various Chromium forks; such patches would contain the code `PP-FP-FONT` as a prefix in the Subject and have a short description and the URL to the wiki
4. we can point to one of the forks as the upstream for such patch; downstreams are free to diverge but then it should be reported in the table with some notes

Pros:
* we have a single place where to check all the context about a patch and the **reason why the patch exists in the first place** (if you see some of the patches I created, I try to put in there the rationale sometimes)
* we can get rid of the privacy patch as soon as it is obsolete or replaced by a new one
* we can do versioning of the patches themselves `PP-FP-FONT-v1` etc
* we standardize the build system so that adding/editing/removing these patches is easier for new developers
* we can create patches for new privacy problems in a coordinated way (use bounty programs etc)

Cons:
* there can be disagreement both on the privacy problem and on the implementation; in such cases the best we could do is to document the differences
* it requires discipline to keep this running and up-to-date
* wikis require lots of manual changes, one would want to automate most of the synchronisation over time
* the Android landscape is quite different from Linux/Windows/Mac (that is why Bromite exists, after all), there will be scope problems in covering all of them

You can keep thinking about this and come up with more ideas and more pros/cons. I am not thinking of them as RFCs but just "handles" for our "privacy problems" and the related solutions we implement.

I can try to give this a bit more shape and start something about it, but let's not forget that these are open source projects where we spend our free time...one Chromium fork might not participate, or not in the way we prefer, and we must be completely fine with that. Incentives play an important role here.

**Reply by wchen342** [reply]
@csagan5 I appreciate the detailed writing about your thoughts. I am not that good at lining out things logically so it is very helpful.

> this is how I can foresee that an improvement from current situation - without creating an absolute centralisation - could be arranged

This is actually what I meant but I wasn't able to express clearly. My intention was not to create an extra project or organization, but find a better way to: (1) let each other know what they are working on, so if two parties has mutual interests they can work on the same thing together instead of working separately on the same thing, and (2) solve each other's problems since one project may have more knowledge on some psecific topics than the other.

The idea of some kind of wiki is great, but it needs to be something easy to maintain I think. The indexing of related issues is one of the exact thing I had in mind when starting this discussion. It can be expanded to some kind of task list so people from different projects can pick one, finish it and share the results.

About where to have conversations, I am not sure whether real time conversation is necessary. There is a matrix room for ungoogled-chromium so if necessary we can create an additional room there. Otherwise probably some form of dedicated issue tracker shall be good enough.

About the disagreements between projects, it can be a problem and some more thought certainly needs to be put into it. I think in most of the cases things can be toggable and because they shall be patches anyway it shouldn't be too hard to exclude/include a certain functionality, but there can be other problems like the canvas noise. I think one way to solve the problem is to document things so changes can be customized.

**Reply by Eloston** [reply]
@csagan5 First off, you did a great job explaining the larger challenges in project management and cross-project collaboration. For example, I particularly like your point about feedback from the vocal minority; it's an issue I've been thinking about on and off for a while.

Most importantly, you've raised an important point that I hadn't considered initially: **We need to clarify intentions/objectives, and establish what we want to share.** We've been focusing on mechanisms of sharing too much. Now, I'll try to clarify my vision for ungoogled-chromium.

From its inception, this project has one primary objective: removing integration with Google web services. All other concerns are secondary. You can think of it as the UNIX philosophy applied to the project vision: Focus on one thing, and do it well.

For a while now, we've basically solved that problem. Granted, there are more technical challenges that we're running into (notably #845 and #1087), but those don't impact much with respect to the primary objective. That's why I have the last and least important objective to "feature tweaks to enhance privacy, control, and transparency." It's really a catch-all for everything else that's nice to have.

The reason why I operate this way is because I don't have a larger vision for ungoogled-chromium beyond the primary objective. When I published this project about six years ago, I published it with the ultimate hope that other developers could use my work to build something larger. Developers like @csagan5, who not only have the knowledge, but also the passion and vision to build larger projects like Bromite (even if it's only incrementally larger). Personally, I am not that invested in the field of privacy; there are many social, political, and technical advancements that I do not keep in touch with. The community around ungoogled-chromium is my main channel for this kind of information. So whenever I encounter those who push for more out of ungoogled-chromium, through their voice or their PRs, I always find it difficult to make decisions.

To bring this back to the idea of privacy enhancements, I don't think ungoogled-chromium will collaborate well with other privacy-focused projects as long as I lead the project. Essentially, I don't have the same commitment or passion to dive into this field compared to others like @csagan5. At the same time, I don't want to completely brush aside any opportunities to help ungoogled-chromium and others grow.

Perhaps we should move into a new discussion (maybe on Issues, because Discussions doesn't have a great layout for long discussions) to consider how we want to handle the third objective of this project (privacy, control, and transparency) going forward. What do you guys think?

**Reply by wchen342** [reply]
> Perhaps we should move into a new discussion

Right, discussion keeps folding following-up comments and that's very inconvenient.

> You can think of it as the UNIX philosophy applied to the project vision: Focus on one thing, and do it well.

Now I start to understand how you made decisions on issues and feature requests. I had always had difficulties when deciding whether some functionalities/enhancements shall be included. From my own view of perspective, there were some ideas I felt interested but also beyond the scope of this project so I had to kill off the requests, and it had always been something I am consufed.

> I don't have the same commitment or passion to dive into this field compared to others like @csagan5. At the same time, I don't want to completely brush aside any opportunities to help ungoogled-chromium and others grow.

Essentially the Android variant has already gone beyond your initial intention I believe, especially the extension-support part (I do explicitly state that the extension version is not officially part of the project though). However, I am inclined to start a new project separately actually, for one very simple reason: `ungoogled-chromium` is one of the best known open source chromium fork, and it has a steady and visible user base. Anything start from scretch will have a hard time gaining users, and a good user base is critical for developments since the more users, the more feedbacks you have which in turn improves the software.
That being said, I still try to stick to the project objective as much as possible (hence the killing off of some ideas), but at some point may be it _will_ be necessary to start a spin-off of some kind.

Also I think it _is_ necessary to look into further topics related to privacy and other enhancement at some time, for the historical (I am not sure whether this is the right word here) reason. I believe @Eloston knows better than me about this, but if we still remember, there used to be three projects: iridium, inox and ungoogled-chromium. The other two merged into this project, but that left a vacuum behind since iridium is supposed to enhance security, and inox probably had similar goals too. Another view point of this is that since the project keeps growing, there will always be more users with more requests, and some of those requests will be focused on improving privacy and security eventually. At some point this will become something inevitable.

**Reply by csagan5** [reply]
> That's why I have the last and least important objective to "feature tweaks to enhance privacy, control, and transparency." It's really a catch-all for everything else that's nice to have.

For Bromite it is actually the opposite: I started the project with the acknowledgement that the Android version of Chromium is the most affected by **user agent limitations** that impair privacy, control and transparency; erosion had begun many years before Bromite creation, pretty much as early as mobile browsers became more important.

> Developers like @csagan5, who not only have the knowledge, but also the passion and vision to build larger projects like Bromite (even if it's only incrementally larger). Personally, I am not that invested in the field of privacy; there are many social, political, and technical advancements that I do not keep in touch with. The community around ungoogled-chromium is my main channel for this kind of information.

I do not think of myself as having a lot of knowledge on this field; I reckon the challenge of keeping up to date, but I am more of a mechanic than an innovator; I think that open source projects like these need contributors more than users: different points of view are a strength, and with more participants there will be also a better knowledge base.

As for Bromite being larger, we need to stand on the shoulders of giants, right? When I started it only ungoogled-chromium was properly maintained and it still is, I am very glad it exists and of the collaborations done so far.

> To bring this back to the idea of privacy enhancements, I don't think ungoogled-chromium will collaborate well with other privacy-focused projects as long as I lead the project. Essentially, I don't have the same commitment or passion to dive into this field compared to others like @csagan5. At the same time, I don't want to completely brush aside any opportunities to help ungoogled-chromium and others grow.
> 

I think you underestimate your role in this case: as you said, you can help fulfilling the goal of ungoogled-chromium following UNIX philosophy to remove integration with Google web services. You have a great advantage there: the scope is easy to identify and you can avoid a lot of the discussions that come attached to privacy, control and transparency.

> Perhaps we should move into a new discussion (maybe on Issues, because Discussions doesn't have a great layout for long discussions) to consider how we want to handle the third objective of this project (privacy, control, and transparency) going forward. What do you guys think?

Fine by me.

> However, I am inclined to start a new project separately actually, for one very simple reason: `ungoogled-chromium` is one of the best known open source chromium fork, and it has a steady and visible user base. Anything start from scratch will have a hard time gaining users, and a good user base is critical for developments since the more users, the more feedbacks you have which in turn improves the software.

It is not fair to "hijack" one open source brand/project for a different project...it would have all the traits of [misleading advertising](https://en.wikipedia.org/wiki/False_advertising); part of your user base will feel misled by these techniques.

> I believe @Eloston knows better than me about this, but if we still remember, there used to be three projects: iridium, inox and ungoogled-chromium. The other two merged into this project, but that left a vacuum behind since iridium is supposed to enhance security, and inox probably had similar goals too.

How did they merge into ungoogled-chromium? I missed the chronology there. I consider Vanadium to be the "spiritual successor" of past security hardening Chromium forks; but @thestinger can probably confirm that none of us currently has all the resources that would be necessary to make any of these forks thrive.

**Reply by thestinger** [reply]
Yeah, we currently don't have substantial resources to assign to Vanadium. We're also largely just having it depend on the GrapheneOS hardening rather than shoving stuff into there which is provided by the OS via libc, etc.

Also very focused on the issues with Copperhead at the moment.

**Reply by wchen342** [reply]
> It is not fair to "hijack" one open source brand/project for a different project...it would have all the traits of misleading advertising; part of your user base will feel misled by these techniques.

Well currently I am still sticking to the objectives of the project as I stated above, the spin-off is just a hypothetical scenario. But still, thanks for the reminder.
Honestly speaking however, after all these time looking around and seeing different codes and approaches, I feel like Bromite is really the best choice for a privacy enhanced version of Android Chromium fork and also kiwi is still living and well, so I am not sure whether the Android project is needed..judging from the downloading counts it has way less users than Bromite anyway.

> How did they merge into ungoogled-chromium?

I think I read that somewhere but I cannot find the source anymore. Maybe @Eloston can explain this a little more cause I am pretty new here despite being active for a while.

**Reply by csagan5** [reply]
> the spin-off is just a hypothetical scenario. But still, thanks for the reminder.

This thread is a wild ride :) we went from trying to coordinate efforts for the privacy patches to creating a new separate Chromium fork project (which technically would make the problem a little bit worse); just a funny note.

> after all these time looking around and seeing different codes and approaches, I feel like Bromite is really the best choice for a privacy enhanced version of Android Chromium fork

Your (and anybody else's) patches are always welcome.

> kiwi is still living and well

Is it still using an old version of Chromium as base? I am glad security fixes are being backported; I see it is developed through a [separate internal repository](https://github.com/kiwibrowser/src/commit/503b2a9351fd52ca96b063d78fb3600e42e8fe74#diff-7d027a8a25b7ec29ff59f99bf1edff910a24c55dab6d8df662d6e31b6f0b894f). One more reason to aim for reproducible builds if you want to call anything "open source".

> I am not sure whether the Android project is needed..judging from the downloading counts it has way less users than Bromite anyway.

To start a project I suggest starting from identifying user needs that you want to match and check whether they are satisfied or not, not from the downloads count; it is a poor metric to gauge user needs. You cannot count downloads of something that does not exist yet. Having said that, what are the features that you miss in Bromite and that are implemented in ungoogled-chromium-android? We can start a separate discussion on that; I have recently opened the [Discussion in Bromite too](https://github.com/bromite/bromite/discussions).

**Reply by wchen342** [reply]
> I see it is developed through a separate internal repository.

Yes that's one thing I do not feel comfortable about, but I guess it's their MO.

> You cannot count downloads of something that does not exist yet.

The android version has been existing for one year and a half, so I won't say that's exactly "not exist".

> what are the features that you miss in Bromite and that are implemented in ungoogled-chromium-android?

For me, it's actually the other way around: I do not miss features. I feel Bromite has too many extra features. There are two reasonings behind this: (1) from a user perspective, trying to avoid Google forcing extra things on me is the reason I do not use chromium in the first place, so I certainly don't want other extra features I don't need. and (2) from a developer perspective, I am not sure all the mid-sized to large modifications are properly tested. No offence, but if they are not unit tested, santinized or fuzzied, then the larger a modification is, the larger the risk of a security breach. I guess the same can be said for some of ungoogled-chromium's patches, but most of the things here is removal. That is also the reason that the only huge modification in uc-android is extension, and I am constantly warning that it is highly experimental and I am pretty sure it is not secure in some way.

Actually I am thinking maybe I shall merge the normal version of uc-android into bromite and only leave the extension part, and apply it on Bromite instead. I will give it a little more thoughts and open a discussion in your repo.

**Reply by csagan5** [reply]
> The android version has been existing for one year and a half, so I won't say that's exactly "not exist".

I thought you were talking about starting a new Android Chromium fork project.

> I feel Bromite has too many extra features [...] I certainly don't want other extra features I don't need

I am not aware of this problem; can you make a list of all the extras? perhaps in Bromite's Discussions

> from a developer perspective, I am not sure all the mid-sized to large modifications are properly tested.

They are not; this is what I was referring to when talking about the fact that all these Chromium forks do not have enough resources.

**Reply by dm17** [reply]
> > Perhaps we need to identify how we will be upstream/downstream and what will be the channel/way to get the news about the updates, and then (3) would be solved.

I imagine it would be structured such that Bromite, Vanadium, and Marmaduke's builds are branches off of ungoogled-chromium. Despite the projects potentially having different goals - it still seems worthwhile to deduplicate work and make sharing their improvements easier.

So basically if each project signs off, then a privacy/security enhancement could be merge into the base ungoogled-chromium. For example:
1) Merge everything each project wants into a base project (ungoogled-chromium)
2) Have each project branch off of the base project with its unique patches
3) Have a review project for each privacy/security enhancement such that it could get merged into the base project

Would that make sense?

And perhaps Marmaduke & the Iridium will want to help out too. As I understand it the only benefit of Marmaduke's ungoogled builds is the addition of some codecs?

**Reply by csagan5** [reply]
> and Marmaduke's builds

No source has ever been released for these, in violation of the spirit of the projects and the open source license.

**Reply by thestinger** [reply]
We'd like to have a shared base for Vanadium but we don't want to remove Google as an option for features like search where it's simply one of several alternatives and anyone else could easily implement the same service. Similarly, we don't really have a problem with functionality that's not enabled by default as long as it doesn't depend on Play services.

We also care a lot about very quickly updating to the latest releases so invasive patches requiring a lot of maintenance are a problem. Having far more complex patches to accomplish things that could be done in a simpler way isn't something we want. If you wanted to have a shared base it would need to be a new repository because the approach here doesn't really work for us.

**Reply by kmk3** [reply]
@thestinger 44 minutes ago:

> We'd like to have a shared base for Vanadium but we don't want to remove
> Google as an option for features like search where it's simply one of several
> alternatives and anyone else could easily implement the same service.
> Similarly, we don't really have a problem with functionality that's not
> enabled by default as long as it doesn't depend on Play services.
> 
> We also care a lot about very quickly updating to the latest releases so
> invasive patches requiring a lot of maintenance are a problem. Having far
> more complex patches to accomplish things that could be done in a simpler way
> isn't something we want. If you wanted to have a shared base it would need to
> be a new repository because the approach here doesn't really work for us.

How about:

* List all patches from ungoogled-chromium (and maybe similar relevant patches
  from the other projects) on a comment (on a new thread or maybe in here)
* Every project maintainer copies and pastes the list on a new comment, with an
  accept/maybe/deny beside each patch

That way, the common denominator would be found.

If it differs too much from what ungoogled-chromium does, it could be done on
another base on which ungoogled-chromium and everyone else would build upon.
Perhaps a "liberated" Chromium that only removes all (client-side) proprietary
software dependencies (such as Play Services)?  It would be nice to have that,
if anything so that distros/roms that currently ship standard Chromium would
replace it with the more libre version (and thus potentially reduce the need to
coordinate patches between distros/roms).

**Reply by dm17** [reply]
Following up to see if there are fresh ideas on how to have an ungoogled-chromium-base without disrupting ungoogled-chromium, vanadium, bromite, etc. It still seems that it could bring a lot of benefit to each project. It could also bring clarity & ease for testing each other's patches. Overall it seems efficient long-term, but I wouldn't want it to complicate the process of building for multiple architectures (android, musl, glibc, etc). Maybe we should draw out an ideal build pipeline to see what the various projects think of inheriting from it?

**Comment by ghost** [comment]
@Eloston  you may can grab the finger print protection code from ddg browser 

Ddg browser are open source as well i think is ok if we want to use it

**Comment by specter78** [comment]
Brave has also done good work in fingerprint protection. Maybe that can be imported in UC.

**Reply by ghost** [reply]
Agree @specter78
